### PR TITLE
Fix tz offset bounds

### PR DESCRIPTION
**Pull request by robertbastian**
Apparently everything up to 24:59:59 is valid posix, even though no time zone has gone there yet.

### CODE REVIEW

File: components/timezone/src/types.rs [hunk 1]
```diff
@@ -27,7 +27,7 @@ impl GmtOffset {
     /// overflows or underflows.
     pub fn try_from_offset_seconds(seconds: i32) -> Result<Self, TimeZoneError> {
         // Valid range is from GMT-12 to GMT+14 in seconds.
```
Comment (Manishearth) [inline components/timezone/src/types.rs:29]: nit: this line needs to be fixed too

File: components/timezone/src/types.rs [hunk 2]
```diff
@@ -93,32 +93,24 @@ impl GmtOffset {
             _ => return Err(TimeZoneError::InvalidOffset),
         };
 
-        let seconds = match chars {
+        let hours = match chars {
+            &[h1, h2] => try_get_time_component([h1, h2]),
```
Comment (Manishearth) [inline components/timezone/src/types.rs:97]: issue: won't this fail to match `[h1, h2, m1, m2]`?

**Review by Manishearth — CHANGES_REQUESTED** [review]
I am very excited for the future in which Kiribati gains territorial dependencies to the east

File: components/timezone/src/types.rs (hunk 2 above)
Comment (robertbastian) [reply components/timezone/src/types.rs:97]: whoops

**Review by Manishearth — DISMISSED** [review]

**Comment by sffc** [comment]
Is this a follow-up to https://github.com/unicode-org/icu4x/pull/2943?

**Comment by robertbastian** [comment]
No, that didn't touch the bounds.

**Comment by nordzilla** [comment]
I want to make sure we do the correct thing here. There are a number of things to consider.

@robertbastian we had a conversation offline regarding the required range and I sent you this [GNU page](https://www.gnu.org/software/libc/manual/html_node/TZ-Variable.html) about TZ Variables, which expresses that the range spans `-24:59:59` to `+24:59:59`

image

However, [RFC 8536](https://www.rfc-editor.org/rfc/rfc8536) for TZif suggests that the POSIX range is `-24:59:59` to `+25:59:59`. Also note that while TZif expresses that values _should_ be within the POSIX range, this format allows the offset to be anywhere in the `i32` range except for the value `-2**31` exactly.

image

I'm trying to find a definite answer about the POSIX range, but I'm not seeing anything, even after looking at the POSIX resources directly, though it's very large and I may be missing something. 

We should also note that, in practice, the [IANA time zone database](https://github.com/eggert/tz/blob/4758e88f99c71af65d73ce730e87438466032a5f/etcetera#L36-L78) does not include any entries outside of the range of GMT-12 to GMT+14.

image

Is there a clear benefit for ICU4X supporting the full POSIX range compared to the IANA range that exists? 

One nice aspect about only supporting the range of `-12` to `+14` is that this value in the ICU4X data can fit into only 2 bytes. Extending the range to the POSIX supported range would increase the data size for each GMT offset when in practice the data won't contain a value outside of `-12` to `+14`. 

If there were to be a future GMT offset that exceeds what we can fit in the `u16` (i.e. beyond `-18` to `+18`), then I think that would be a large enough change in the world that might warrant making a `V2` variant of those data structs. 

This opens the question about whether we should be consistent across all of ICU4X, both in datagen and at runtime. 

We could reasonably restrict ICU4X data to the `-12` to `+14` range (since that's what currently exists) but allow the full POSIX range at runtime. Or would it be better to be consistent across the board?

In any case, I'd like to document our decision clearly in the TZDB Datagen PR that is currently in review. 

@sffc @Manishearth @samchen61661 @robertbastian 

Thoughts?

**Comment by robertbastian** [comment]
Thanks for the detail. I think current time zones are a risky assumption, if some island changed to -13 tomorrow, which is not totally unprecedented, we shouldn't have to rewrite our code. We could, of course, but it'd be nice if it just kept working and we can focus on all the other code that such an event would break.

I'm also happy with +-18 for efficiency, those offsets are much less likely.

**Comment by sffc** [comment]
I see no reason to support anything in excess of `±18`

Note that we can't really fit this in an i16; we need one more bit unfortunately. However, `±18` fits in 17 bits (u16 and one additional sign bit), so an array of them would be easy to bit-pack.

**Comment by robertbastian** [comment]
Are we happy to submit like this, or should I use tighter bounds?

**Comment by sffc** [comment]
±18 is fine with me

**Comment by Manishearth** [comment]
Deferring to the others on this

**Review by nordzilla — APPROVED** [review]
Happy with `±18`. Thanks!

**Review by sffc — COMMENTED** [review]
lgtm

**Pull request merged**